dialogue-clients/src/main/java/com/palantir/dialogue/core/ChannelCache.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.palantir.conjure.java.api.config.service.ServiceConfiguration;
+import com.palantir.conjure.java.client.config.ClientConfiguration;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.hc4.ApacheHttpClientChannels;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.Safe;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.concurrent.ThreadSafe;
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ThreadSafe
+final class ChannelCache {
+    private static final Logger log = LoggerFactory.getLogger(ChannelCache.class);
+
+    // TODO(dfox): consider making this caffeine too?
+    private final Map<String, ApacheCacheEntry> apacheCache = new ConcurrentHashMap<>();
+    private final LoadingCache<ChannelCacheKey, Channel> channelCache = Caffeine.newBuilder()
+            .maximumSize(500) // I don't think it's reasonable to need 500 clients at the same time
```

Comment:
Not sure I'd bound this since the expensive resource is likely the client anyhow.

Reply:
What happens if we don't cache enough of these? I suppose we recreate them, which is probably fine, but it potentially results in losing connection state (balanced scores, pin-until-error state, queues, etc).

Maybe we want weak values so we don't hold onto channels that are no longer in use? Not a huge fan of the weak reference model either...

It's probably fine as is. Perhaps we should instrument the cache?

Reply:
Yeah this is kinda what i was thinking with this stuff below
```java
        if (count > 400) {
            log.warn("ChannelCache nearing capacity - possible bug?", SafeArg.of("count", count));
        }
```